### PR TITLE
card backend implementation

### PR DESCRIPTION
**Pull request by majdantaki**

### CODE REVIEW

File: kanbananza-backend/package.json [hunk 1]
```diff
@@ -40,6 +40,7 @@
     "express-jwt": "^6.0.0",
     "express-session": "^1.17.1",
     "jsonwebtoken": "^8.5.1",
+    "lint": "^0.7.0",
```
Comment (loucadufault) [inline kanbananza-backend/package.json:43]: ?

File: kanbananza-backend/src/main/DTO/card.js [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+  }
+
+  static fromDocument(card) {
+    return new this({
+      id: card._id,
+      label: card.label,
+      column: card.columnId,
+    });
+  }
+}
```
Comment (loucadufault) [inline kanbananza-backend/src/main/DTO/card.js:15]: we moved this to the model classes, check the latest code

File: kanbananza-backend/src/main/models/card.js [hunk 3]
```diff
@@ -0,0 +1,11 @@
+import mongoose from "mongoose";
+
+const Schema = mongoose.Schema;
+
+const cardSchema = new Schema({
+  label: { type: String, required: true, maxLength: 100 },
+  column: { type: Schema.ObjectId, required: true, ref: "Column" },
```
Comment (loucadufault) [inline kanbananza-backend/src/main/models/card.js:7]: name:
columnId:

File: kanbananza-backend/src/main/services/card.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    if (e instanceof Error.ValidationError) {
+      throw new HttpError({
+        code: 400,
+        message: "Invalid field(s)",
+        body: Object.values(e.errors).map((error) => error.message),
+      });
+    } else {
+      throw e;
+    }
+  }
```
Comment (loucadufault) [inline kanbananza-backend/src/main/services/card.js:23]: this was changed, please check how its done in the other service/controllers

**Review by loucadufault — CHANGES_REQUESTED** [review]

File: kanbananza-backend/src/main/DTO/card.js (hunk 2 above)
Comment (majdantaki) [reply kanbananza-backend/src/main/DTO/card.js:15]: Where is the latest code? not on main?

Comment (majdantaki) [reply kanbananza-backend/src/main/DTO/card.js:15]: oh nvm

File: kanbananza-backend/package.json (hunk 1 above)
Comment (majdantaki) [reply kanbananza-backend/package.json:43]: i reverted it

File: kanbananza-backend/src/main/routes/card.js [hunk 5]
```diff
@@ -0,0 +1,9 @@
+import cardController from "../controllers/card";
+import { validateSchema } from "../middleware/schema_validator";
+import idParamValidator from "../middleware/id_param_validator";
+
+export default (router) => {
+  router.post("/card", validateSchema("createCard"), cardController.create);
+  router.get("/columns", cardController.select);
+  router.get("/column/:id", idParamValidator, cardController.index);
```
Comment (loucadufault) [inline kanbananza-backend/src/main/routes/card.js:8]: /cards
/card/:id

File: kanbananza-backend/src/main/services/card.js [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+  try {
+    return await Card.create({ name, columnId, order });
+  } catch (e) {
+    if (e instanceof Error.ValidationError) {
+      throw new HttpError({
+        code: 400,
+        message: "Invalid field(s)",
+        body: Object.values(e.errors).map((error) => error.message),
+      });
```
Comment (loucadufault) [inline kanbananza-backend/src/main/services/card.js:19]: this should throw ValidationErrors that are caught and attached to HttpError in controller

Comment (loucadufault) [reply kanbananza-backend/src/main/services/card.js:19]: you can check the other services

File: kanbananza-backend/src/main/routes/card.js (hunk 5 above)
Comment (majdantaki) [reply kanbananza-backend/src/main/routes/card.js:8]: oops done

File: kanbananza-backend/src/main/services/card.js [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+  return Card.find().exec();
+};
+
+const findCardById = async (id) => {
+  if (!(await Card.exists({ _id: id }))) {
+    throw new HttpError({
+      code: 404,
+      message: `Card with id ${id} does not exist.`,
+    });
+  }
```
Comment (loucadufault) [inline kanbananza-backend/src/main/services/card.js:26]: delete, since this is handled by controller

File: kanbananza-backend/src/main/services/card.js [hunk 8]
```diff
@@ -0,0 +1,30 @@
+import Card from "../models/card";
+import HttpError from "../http_error";
+import ValidationError from "../validation_error";
+
+const createCard = async ({ name, columnId, order }) => {
+  if (await Card.exists({ columnId, name })) {
+    throw new ValidationError({
+      path: "name",
+      reason: "card name already in use for column.",
```
Comment (loucadufault) [inline kanbananza-backend/src/main/services/card.js:9]: nit: remove "." at end of reason string

**Review by loucadufault — APPROVED** [review]
lgtm except a few things

File: kanbananza-backend/src/main/services/card.js (hunk 8 above)
Comment (majdantaki) [reply kanbananza-backend/src/main/services/card.js:9]: lol!

**Review by loucadufault — APPROVED** [review]
lgtm!

**Comment by RemiCarriere** [comment]
Merging after configuring router and manually testing,

File: kanbananza-backend/src/main/controllers/user.js [hunk 9]
```diff
@@ -24,13 +24,6 @@ const create = async (req, res, next) => {
         })
       );
     }
-
-    if (e instanceof mongoose.Error.ValidationError) {
-      return next(
-        HttpError.fromMongooseValidationError(e, "Invalid user information")
-      );
-    }
-
```
Comment (loucadufault) [inline kanbananza-backend/src/main/controllers/user.js:33]: @RemiCarriere why was this removed? I think saving the user doocument in the service can still throw mongse validation errors that are not checked so far
